### PR TITLE
chore(azurerm): upgrade to 3.3.0

### LINKED ISSUE #145: addon_profile has been deprecated
This module is still using `addon_profile` that is being deprecated in 3.0 AzureRM provider. We receive this warning message when using 2.97.0 azurerm provider and running `terraform plan` or `terraform apply`

>│ Warning: Argument is deprecated
>│ 
>│   with module.aks.azurerm_kubernetes_cluster.main,
>│   on .terraform/modules/aks/main.tf line 10, in resource "azurerm_kubernetes_cluster" "main":
>│   10: resource "azurerm_kubernetes_cluster" "main" {
>│ 
>│ `addon_profile` block has been deprecated and will be removed in version 3.0 of the AzureRM Provider. All properties within the >block will move to the top level.
>│

### LINKED ISSUE #156: Upgrade to AzureRM 3.3.0
AzureRM 3.0 has been out there for some time now. In my eyes it's time to upgrade this module to use AzureRM 3.3.0

The upgrade guide clearly states what needs to be changed:

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-upgrade-guide#resource-azurerm_kubernetes_cluster

### PR DESCRIPTION
**Pull request by eyenx**
Fixes #145
Fixes #156

Changes proposed in the pull request:

Upgrade azurerm to 3.3.0 following the upgrade guide https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-upgrade-guide#resource-azurerm_kubernetes_cluster

Tests succesful:

```
TestTerraformBasicExample 2022-04-22T11:25:27Z command.go:96: Destroy complete! Resources: 18 destroyed.
--- PASS: TestTerraformBasicExample (1473.05s)
PASS
ok      terraform-azurerm-aks/test      1473.061s
```

---

Signed-off-by: Toni Tauro <toni.tauro@adfinis.com>

### CODE REVIEW

**Comment by lonegunmanb** [comment]
Hello @eyenx , thanks for opening this pr. Upgrading to 3.x may introduce some breaking changes, while we're planning on adding a new CI pipeline to this module so the further pr can be tested in GitHub action. We're working on this so the next release won't change module's code dramatically. I'm sorry to say that we might need more time to merge this pr.

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
When apply test/fixture folder, the following errors occured:

```text
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 37:
│   37: output "test_client_key" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 41:
│   41: output "test_client_certificate" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 45:
│   45: output "test_cluster_ca_certificate" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 49:
│   49: output "test_host" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 53:
│   53: output "test_username" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
```

**Comment by lonegunmanb** [comment]
Hi @eyenx , after discussed with my team, we decided to accept this pr before we introduce pipeline to this module. There're some suggestions remain, would you please take a look? Thanks.

File: variables.tf [hunk 1]
```diff
@@ -301,15 +301,16 @@ variable "ingress_application_gateway_subnet_id" {
...
   description = "(Optional) The type of identity used for the managed cluster. Conflict with `client_id` and `client_secret`. Possible values are `SystemAssigned` and `UserAssigned`. If `UserAssigned` is set, a `user_assigned_identity_id` must be set as well."
   type        = string
   default     = "SystemAssigned"
 }
 
-variable "user_assigned_identity_id" {
-  description = "(Optional) The ID of a user assigned identity."
-  type        = string
+variable "identity_ids" {
+  description = "(Optional) The IDs of a user assigned identity."
```
[the remark this answers is not in the document]
Comment (eyenx) [reply variables.tf:312]: Can do that

File: main.tf [hunk 2]
```diff
@@ -80,60 +80,49 @@ resource "azurerm_kubernetes_cluster" "main" {
...
+  azure_policy_enabled = var.enable_azure_policy
 
-    azure_policy {
-      enabled = var.enable_azure_policy
-    }
-
-    oms_agent {
-      enabled                    = var.enable_log_analytics_workspace
+  dynamic "oms_agent" {
+    for_each = var.enable_log_analytics_workspace ? ["oms_agent"] : []
```
[the remark this answers is not in the document]
Comment (eyenx) [reply main.tf:93]: I did the same as on [line 80](https://github.com/Azure/terraform-azurerm-aks/pull/157/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR80) with identity. As far as I can see we do not check there either if it's null or false

File: main.tf [hunk 3]
```diff
@@ -80,60 +80,49 @@ resource "azurerm_kubernetes_cluster" "main" {
...
-      for_each = var.enable_ingress_application_gateway == null ? [] : ["ingress_application_gateway"]
-      content {
-        enabled      = var.enable_ingress_application_gateway
-        gateway_id   = var.ingress_application_gateway_id
-        gateway_name = var.ingress_application_gateway_name
-        subnet_cidr  = var.ingress_application_gateway_subnet_cidr
-        subnet_id    = var.ingress_application_gateway_subnet_id
-      }
+  dynamic "ingress_application_gateway" {
+    for_each = var.enable_ingress_application_gateway == null ? [] : ["ingress_application_gateway"]
```
[the remark this answers is not in the document]
Comment (eyenx) [reply main.tf:100]: I did the same as on [line 80](https://github.com/Azure/terraform-azurerm-aks/pull/157/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR80) with identity. As far as I can see we do not check there either if it's null or false

File: outputs.tf [hunk 4]
```diff
@@ -45,7 +45,7 @@ output "kube_admin_config_raw" {
 }
 
 output "http_application_routing_zone_name" {
-  value = length(azurerm_kubernetes_cluster.main.addon_profile) > 0 && length(azurerm_kubernetes_cluster.main.addon_profile[0].http_application_routing) > 0 ? azurerm_kubernetes_cluster.main.addon_profile[0].http_application_routing[0].http_application_routing_zone_name : ""
+  value = length(azurerm_kubernetes_cluster.main) > 0 && azurerm_kubernetes_cluster.main.http_application_routing_enabled ? azurerm_kubernetes_cluster.main.http_application_routing_zone_name : ""
```
[the remark this answers is not in the document]
Comment (eyenx) [reply outputs.tf:48]: You are right we could get rid of that 

```suggestion
  value = azurerm_kubernetes_cluster.main.http_application_routing_enabled ? azurerm_kubernetes_cluster.main.http_application_routing_zone_name : ""
```

File: variables.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (eyenx) [reply variables.tf:312]:
```suggestion
  description = "(Optional) Specifies a list of User Assigned Managed Identity IDs to be assigned to this Kubernetes Cluster."
```

**Comment by lonegunmanb** [comment]
@eyenx thanks for updating, after run the test I've got the following error:

```text
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 37:
│   37: output "test_client_key" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 41:
│   41: output "test_client_certificate" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 45:
│   45: output "test_cluster_ca_certificate" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 49:
│   49: output "test_host" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 53:
│   53: output "test_username" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
╷
│ Error: Output refers to sensitive values
│
│   on outputs.tf line 57:
│   57: output "test_password" {
│
│ To reduce the risk of accidentally exporting sensitive data that was intended to be only internal, Terraform requires that any root module output     
│ containing sensitive data be explicitly marked as sensitive, to confirm your intent.
│
│ If you do intend to export this data, annotate the output value as sensitive by adding the following argument:
│     sensitive = true
╵
```

I think it's ok to leave the variables' validation blank for now since the existing variables are lacking of proper validation, I'll deal with them in the later pr. Would you please update the `outputs.tf` in the `test/fixture` folder to make the test green light? Thanks.

**Comment by eyenx** [comment]
Done! @lonegunmanb

**Comment by lonegunmanb** [comment]
@eyenx  apology for the confusion, but these errors are complaining that the outputs in `test/fixture` folder are lacking of `sensitive` argument. Would you please correct these issues? Thanks.

**Comment by eyenx** [comment]
I see should I revert my changes then?

**Comment by eyenx** [comment]
@lonegunmanb can you retry? my tests worked now.

**Review by lonegunmanb — APPROVED** [review]

**Comment by lonegunmanb** [comment]
Thanks @eyenx, LGTM!

**Comment by davidkarlsen** [comment]
@lonegunmanb Awesome! Will there be a release soon?

**Comment by lonegunmanb** [comment]
> @lonegunmanb Awesome! Will there be a release soon?

Hi @davidkarlsen , since this release will introduce breaking changes, I'd like to submit several more pull requests which contain breaking changes to improve the current code so the users will experience one time breaking change then everything will be stable.

**Comment by hairmare** [comment]
Hi @lonegunmanb THanks for merging this. Is there any chance that it could get released soon?

**Pull request merged**